src/lib.rs

```diff
@@ -632,14 +723,64 @@ where
     }
 }
 
+#[derive(Default)]
 struct S3Listing {
-    count: i64,
     continuation: Option<String>,
-    objects: Vec<S3Obj>,
+    contents: Vec<S3Object>,
+    common_prefixes: Vec<String>,
+}
+
+#[derive(Debug, Clone)]
+pub enum S3ListingItem {
+    S3Object(S3Object),
+    S3CommonPrefix(String),
+}
+
+impl S3ListingItem {
+    fn object(o: crate::S3Object) -> S3ListingItem {
+        S3ListingItem::S3Object(o)
+    }
+    fn common_prefix(cp: String) -> S3ListingItem {
+        S3ListingItem::S3CommonPrefix(cp)
+    }
+    fn prefix(&self) -> String {
+        match self {
+            S3ListingItem::S3Object(o) => o.key.clone(),
+            S3ListingItem::S3CommonPrefix(cp) => cp.clone(),
+        }
+    }
+    fn unwrap_object(self) -> crate::S3Object {
+        match self {
+            S3ListingItem::S3Object(o) => o,
+            S3ListingItem::S3CommonPrefix(_cp) => panic!("invalid type"),
+        }
+    }
+}
+
+impl S3Listing {
+    fn combined(self) -> Vec<S3ListingItem> {
+        let common_prefixes: Vec<S3ListingItem> = self
+            .common_prefixes
+            .into_iter()
+            .map(S3ListingItem::common_prefix)
+            .collect();
+        let contents: Vec<S3ListingItem> = self
+            .contents
+            .into_iter()
+            .map(S3ListingItem::object)
+            .collect();
+        [common_prefixes, contents].concat()
+    }
```

Comment:
maybe?
```suggestion
        let common_prefixes = self
            .common_prefixes
            .into_iter()
            .map(S3ListingItem::common_prefix);
        let contents = self.contents.into_iter().map(S3ListingItem::object);
        common_prefixes.chain(contents).collect()
```

Reply:
lgtm, I knew there had to be a better way